### PR TITLE
Check for function-like rule and configuration

### PR DESCRIPTION
**Pull request by andreiburdusa**
---

###### Reviewer checklist

- [ ] Test coverage: `stack test --coverage`
- [ ] Public API documentation: `stack haddock`

### CODE REVIEW

**Comment by ttuegel** [comment]
I just pushed fixes for all of the test failures except the `\ceil` ones. What did we decide to do about `\ceil` simplification rules again?

**Comment by traiansf** [comment]
@ttuegel please take a look.

File: kore/src/Kore/Step/Axiom/Evaluate.hs [hunk 1]
```diff
@@ -60,11 +61,13 @@ import Kore.Variables.Fresh
 evaluateAxioms
     :: forall variable simplifier
     .  (SimplifierVariable variable, MonadSimplify simplifier)
-    => [EqualityRule Variable]
+    => StepContext
```
Comment (ttuegel) [inline kore/src/Kore/Step/Axiom/Evaluate.hs]: I'm skeptical about this; `evaluateAxioms` can only be used in a function or simplification context because its arguments are `EqualityRule`s.

File: kore/src/Kore/Step/Rule.hs [hunk 2]
```diff
@@ -777,3 +780,48 @@ substitute subst rulePattern' =
...
+  | AllPathContext
+  deriving (Show, Eq)
+
+
+{- | Check if the term and the 'left' fields of the rules are function-like
+
+-}
+
+checkFunctionLike
+    ::  (Debug variable
```
Comment (ttuegel) [inline kore/src/Kore/Step/Rule.hs]: I suggest `(InternalVariable variable, InternalVariable variable') => ` here instead. Does it need `FreshVariable`?

File: kore/src/Kore/Step/Rule.hs [hunk 3]
```diff
@@ -777,3 +780,48 @@ substitute subst rulePattern' =
...
+        , FreshVariable variable
+        , SortedVariable variable
+        , Show variable
+        , Unparse variable
+        , Debug variable'
+        , FreshVariable variable'
+        , SortedVariable variable'
+        , Show variable'
+        , Unparse variable')
+    => [RulePattern variable']
```
Comment (ttuegel) [inline kore/src/Kore/Step/Rule.hs]: There is nothing here which indicates that this should be the list of _applied_ rules, yet that seems to be critical.

File: kore/src/Kore/Step/Rule.hs [hunk 4]
```diff
@@ -777,3 +780,48 @@ substitute subst rulePattern' =
...
+    -> TermLike variable
+    -> Either String ()
+checkFunctionLike [] _ = pure ()
+checkFunctionLike rules term
+  | TermLike.isFunctionPattern term =
```
Comment (ttuegel) [inline kore/src/Kore/Step/Rule.hs]: Why is it OK to omit this check if we did not apply any rules? There would still be a remainder pattern, right?

File: kore/src/Kore/Step/Step.hs [hunk 5]
```diff
@@ -397,19 +397,42 @@ finalizeRulesParallel
...
-    remainders' <- Monad.Unify.gather $ applyRemainder initial remainder
-    return Step.Results
-        { results = Seq.fromList results
-        , remainders =
-            OrPattern.fromPatterns
-            $ Pattern.mapVariables Target.unwrapVariable <$> remainders'
-        }
+finalizeRulesParallel initial unifiedRules =
+    case Rule.checkFunctionLike (term <$> unifiedRules) (term initial) of
+        Left err -> error $ "Rules:\n" <> show (unparseToString . TermLike.mapVariables Target.unwrapVariable . Rule.left . term <$> unifiedRules) <> err
```
Comment (ttuegel) [inline kore/src/Kore/Step/Step.hs]: Because this case only throws an error, it would be better if the `Right` case did `return ()`  and the rest of the function followed at the previous indentation level.

File: kore/src/Kore/Step/Step.hs [hunk 6]
```diff
@@ -397,19 +397,42 @@ finalizeRulesParallel
...
+    ::  forall unifier variable
+    .   ( SimplifierVariable variable
+        , MonadUnify unifier
+        , Log.WithLog Log.LogMessage unifier
+        )
+    => Rule.StepContext
+    -> Pattern (Target variable)
+    -> [UnifiedRule (Target variable)]
+    -> unifier (Results variable)
+finalizeRulesSequenceWithCheck Rule.SimplificationContext initial unifiedRules =
```
Comment (ttuegel) [inline kore/src/Kore/Step/Step.hs]: I don't think we should assume that the simplification context will always apply rules in sequence.

File: kore/src/Kore/Step/Step.hs [hunk 7]
```diff
@@ -565,8 +589,11 @@ applyRewriteRulesParallel
     -> Pattern variable
     -- ^ Configuration being rewritten
     -> unifier (Results variable)
-applyRewriteRulesParallel unificationProcedure rewriteRules =
-    applyRulesParallel unificationProcedure (getRewriteRule <$> rewriteRules)
+applyRewriteRulesParallel unificationProcedure rewriteRules initialConfig =
+    applyRulesParallel
+        unificationProcedure
+        (getRewriteRule <$> rewriteRules)
+        initialConfig
```
Comment (ttuegel) [inline kore/src/Kore/Step/Step.hs]: Why do we need to specify `initialConfig` explicitly now?

File: kore/src/Kore/Step/Step.hs [hunk 8]
```diff
@@ -397,19 +397,42 @@ finalizeRulesParallel
...
-        remainder = Predicate.fromPredicate (Remainder.remainder' unifications)
-    remainders' <- Monad.Unify.gather $ applyRemainder initial remainder
-    return Step.Results
-        { results = Seq.fromList results
-        , remainders =
-            OrPattern.fromPatterns
-            $ Pattern.mapVariables Target.unwrapVariable <$> remainders'
-        }
+finalizeRulesParallel initial unifiedRules =
+    case Rule.checkFunctionLike (term <$> unifiedRules) (term initial) of
```
Comment (ttuegel) [inline kore/src/Kore/Step/Step.hs]: The list of unified rules is available in the `Results`, so I don't see why this check shouldn't happen in the caller.

File: kore/src/Kore/Step/Axiom/Evaluate.hs (hunk 1 above)
Comment (traiansf) [reply kore/src/Kore/Step/Axiom/Evaluate.hs]: True, but we still need to distinguish between the two cases (Function evaluation vs. simplification).

File: kore/src/Kore/Step/Rule.hs (hunk 4 above)
Comment (traiansf) [reply kore/src/Kore/Step/Rule.hs]: I think that if we don't apply any rules then there is no change to the original pattern, so, in principle, the remainder should be itself.

File: kore/src/Kore/Step/Step.hs [hunk 9]
```diff
@@ -397,19 +397,42 @@ finalizeRulesParallel
...
+        )
+    => Rule.StepContext
+    -> Pattern (Target variable)
+    -> [UnifiedRule (Target variable)]
+    -> unifier (Results variable)
+finalizeRulesSequenceWithCheck Rule.SimplificationContext initial unifiedRules =
+    finalizeRulesSequence initial unifiedRules
+finalizeRulesSequenceWithCheck _ initial unifiedRules =
+    case Rule.checkFunctionLike (term <$> unifiedRules) (term initial) of
+        Left err -> error $ "Rules:\n" <> show (unparseToString . TermLike.mapVariables Target.unwrapVariable . Rule.left . term <$> unifiedRules) <> err
```
Comment (virgil-serbanuta) [inline kore/src/Kore/Step/Step.hs]: Line length

File: kore/src/Kore/Step/Axiom/Evaluate.hs (hunk 1 above)
Comment (virgil-serbanuta) [reply kore/src/Kore/Step/Axiom/Evaluate.hs]: Yes, but why do we distinguish by using a context, as opposed to saying what we want? Why is it better to tell the lower levels "You are doing OnePath stuff right now, so do whatever's right for that" as opposed to "You should only accept function-like stuff when applying rules"?

If you do the second, then the higher-level code knows about high-level concepts like OnePath, function evaluation, All Path and similar stuff, while the lower level code knows about lower level concepts like "we can only handle function-like stuff".

File: kore/src/Kore/Step/Axiom/Evaluate.hs [hunk 10]
```diff
@@ -34,45 +43,58 @@ import Kore.Step.Remainder
...
+-- ^ TODO (andrei.burdusa): remove this
+
+evaluateAxioms'
+    :: forall variable simplifier
+    .  (SimplifierVariable variable, MonadSimplify simplifier)
+    => StepContext
+    -> [EqualityRule Variable]
+    -> TermLike variable
+    -> Syntax.Predicate variable
+    -> simplifier (Result.Results (Step.UnifiedRule (Target variable)) (Pattern variable))
```
Comment (ttuegel) [inline kore/src/Kore/Step/Axiom/Evaluate.hs]:
```suggestion
    -> simplifier (Step.Results variable)
```

File: kore/src/Kore/Step/Axiom/Evaluate.hs [hunk 11]
```diff
@@ -34,45 +43,58 @@ import Kore.Step.Remainder
...
+evaluateAxioms'
+    context
     definitionRules
     patt
     predicate
   | any ruleIsConcrete definitionRules
   , not (TermLike.isConcrete patt)
-  = return AttemptedAxiom.NotApplicable
+  = return $ Result.Results mempty mempty
```
Comment (ttuegel) [inline kore/src/Kore/Step/Axiom/Evaluate.hs]:
```suggestion
  = return mempty
```

File: kore/src/Kore/Step/Axiom/Evaluate.hs [hunk 12]
```diff
@@ -95,23 +117,31 @@ evaluateAxioms
         keepResultUnchanged = id
         introduceDefinedness = flip Pattern.andCondition
         markRemainderEvaluated = fmap TermLike.mkEvaluated
+        rules = Result.appliedRule <$> Result.results result
 
     simplifiedResults <-
-        OrPattern.simplifyPredicatesWithSmt
-           predicate (Step.gatherResults result)
+        traverse (OrPattern.simplifyPredicatesWithSmt predicate)
```
Comment (ttuegel) [inline kore/src/Kore/Step/Axiom/Evaluate.hs]: I suggest using a `Traversal` here, so that you don't need to zip the results back up with the rules later:
```suggestion
        Lens.traverseOf (field @"results" . Lens.mapped . field @"result") 
            (OrPattern.simplifyPredicatesWithSmt predicate)
            result
```

File: kore/src/Kore/Step/Axiom/Evaluate.hs [hunk 13]
```diff
@@ -95,23 +117,31 @@ evaluateAxioms
...
+            ((fmap Result.result . Result.results) result)
 
     simplifiedRemainders <-
```
Comment (ttuegel) [inline kore/src/Kore/Step/Axiom/Evaluate.hs]: I know you didn't change this part of the code, but the same comment applies as just above: 
```suggestion
    simplifiedRemainders <- Lens.traverseOf (field @"remainders") (OrPattern.simplifyPredicatesWithSmt predicate) simplifiedResults
```

File: kore/src/Kore/Step/Axiom/Evaluate.hs [hunk 14]
```diff
@@ -34,45 +43,58 @@ import Kore.Step.Remainder
...
     definitionRules
     patt
     predicate
   | any ruleIsConcrete definitionRules
   , not (TermLike.isConcrete patt)
-  = return AttemptedAxiom.NotApplicable
+  = return $ Result.Results mempty mempty
   | otherwise
-  = AttemptedAxiom.maybeNotApplicable $ do
+  = maybeResults $ do
```
Comment (ttuegel) [inline kore/src/Kore/Step/Axiom/Evaluate.hs]: We don't need this, or the `guard` below, anymore.

```suggestion
  = do
```

File: kore/src/Kore/Step/Axiom/Evaluate.hs [hunk 15]
```diff
@@ -124,8 +154,11 @@ evaluateAxioms
     rejectNarrowing (Result.results -> results) =
         (Monad.guard . not) (Foldable.any Step.isNarrowingResult results)
 
-    applyRules initial rules =
-        Monad.Unify.maybeUnifierT
+    applyRules
+        (Step.toConfigurationVariables -> initial)
+        rules
+      = Monad.Unify.maybeUnifierT
+        $ Step.checkFunctionLikeResults context initial
```
Comment (ttuegel) [inline kore/src/Kore/Step/Axiom/Evaluate.hs]: I still want you to move this check into `Kore.Step.Axiom.EvaluationStrategy`.

File: kore/src/Kore/Step/Step.hs [hunk 16]
```diff
@@ -400,9 +401,12 @@ finalizeRulesParallel
 finalizeRulesParallel initial unifiedRules = do
     let initialCondition = Conditional.withoutTerm initial
     results <-
-        Foldable.fold <$> traverse (finalizeRule initialCondition) unifiedRules
-    let unifications = MultiOr.make (Conditional.withoutTerm <$> unifiedRules)
-        remainder = Predicate.fromPredicate (Remainder.remainder' unifications)
+        Foldable.fold
```
Comment (ttuegel) [inline kore/src/Kore/Step/Step.hs]: Why did this change?

File: kore/src/Kore/Step/Step.hs [hunk 17]
```diff
@@ -535,17 +585,16 @@ applyRulesParallel
...
     -> unifier (Results variable)
 applyRulesParallel
     unificationProcedure
     -- Wrap the rule and configuration so that unification prefers to substitute
     -- axiom variables.
     (map toAxiomVariables -> rules)
-    (toConfigurationVariables -> initial)
-  =
-    unifyRules unificationProcedure initial rules
+    initial
```
Comment (ttuegel) [inline kore/src/Kore/Step/Step.hs]: Why did we remove `toConfigurationVariables` here?

File: kore/src/Kore/Step/Step.hs [hunk 18]
```diff
@@ -580,19 +637,18 @@ applyRulesSequence
...
-    -> Pattern variable
+    -> Pattern (Target variable)
     -- ^ Configuration being rewritten
     -> [RulePattern variable]
     -- ^ Rewrite rules
     -> unifier (Results variable)
 applyRulesSequence
     unificationProcedure
-    -- Wrap the rule and configuration so that unification prefers to substitute
+    initial
```
Comment (ttuegel) [inline kore/src/Kore/Step/Step.hs]: Why did we remove `toConfigurationVariables` here?

File: kore/test/Test/Kore/Step/Axiom/Evaluate.hs [hunk 19]
```diff
@@ -54,7 +55,10 @@ test_evaluateAxioms =
     , doesn'tApply "F(x) => G(x) [concrete] doesn't apply to F(x)"
         [axiom_ (f x) (g x) & concreteEqualityRule]
         (f x, makeTruePredicate)
-    , applies "F(x) => G(x) [concrete] apply to F(a)"
+    , doesn'tApply "F(x) => G(x) doesn't apply to F(top)"
+        [axiom_ (f x) (g x)]
+        (f mkTop_, makeTruePredicate)
+     , applies "F(x) => G(x) [concrete] apply to F(a)"
```
Comment (ttuegel) [inline kore/test/Test/Kore/Step/Axiom/Evaluate.hs]:
```suggestion
    , applies "F(x) => G(x) [concrete] apply to F(a)"
```

File: kore/test/Test/Kore/Step/Simplification/Integration.hs [hunk 20]
```diff
@@ -321,6 +322,19 @@ test_simplificationIntegration =
...
 
+simplificationRulePattern
+    :: InternalVariable variable
+    => TermLike variable
+    -> TermLike variable
+    -> RulePattern variable
+simplificationRulePattern left right =
+    patt
+        { attributes = (attributes patt)
+            { Attribute.simplification = Simplification True}
```
Comment (ttuegel) [inline kore/test/Test/Kore/Step/Simplification/Integration.hs]:
```suggestion
            { Attribute.simplification = Simplification True }
```

File: kore/test/Test/Kore/Step/Simplification/Integration.hs [hunk 21]
```diff
@@ -321,6 +322,19 @@ test_simplificationIntegration =
         assertEqual "" expect actual
     ]
 
+simplificationRulePattern
+    :: InternalVariable variable
+    => TermLike variable
+    -> TermLike variable
+    -> RulePattern variable
+simplificationRulePattern left right =
+    patt
```
Comment (ttuegel) [inline kore/test/Test/Kore/Step/Simplification/Integration.hs]: I suggest using a lens here:
```suggestion
    patt & Lens.set (field @"attributes" . field @"simplification") (Simplification True)
```

File: kore/test/Test/Kore/Step/Step.hs [hunk 22]
```diff
@@ -224,20 +245,23 @@ test_applyRewriteRule_ =
     [ testCase "apply identity axiom" $ do
         let expect = Right [ OrPattern.fromPatterns [initial] ]
             initial = pure (mkElemVar Mock.x)
-        actual <- applyRewriteRule_ initial axiomId
+        actual <- applyRewriteRule_ applyRewriteRulesParallel initial axiomId
```
Comment (ttuegel) [inline kore/test/Test/Kore/Step/Step.hs]: Why did we make this change? We don't seem to call `applyRewriteRule_` with any argument other than `applyRewriteRulesParallel`.

File: src/main/kore/prelude.kore [hunk 23]
```diff
@@ -33,7 +33,7 @@ module INJ
     \equals{S3,R}(
       inj{S2,S3}(inj{S1,S2}(T:S1)),
       inj{S1,S3}(T:S1))
-  []
+  [simplification{}()]
```
Comment (ttuegel) [inline src/main/kore/prelude.kore]: Why was this necessary?

File: src/main/k/working/tests/fresh-vars/fresh-vars.k [hunk 24]
```diff
@@ -8,6 +8,6 @@ module FRESH-VARS
     requires N >Int 0
   rule foo(0) => .Seq
 
-  syntax X ::= "foo"
-  rule <k> foo => foo(3) </k>
+  syntax KItem ::= "foo"
```
Comment (ttuegel) [inline src/main/k/working/tests/fresh-vars/fresh-vars.k]: Why was this necessary?

File: kore/src/Kore/Step/Step.hs (hunk 16 above)
Comment (andreiburdusa) [reply kore/src/Kore/Step/Step.hs]: Honestly, I'm not sure. It wasn't exceeding 80 characters. I'll change it back

File: kore/src/Kore/Step/Step.hs (hunk 17 above)
Comment (andreiburdusa) [reply kore/src/Kore/Step/Step.hs]: I think Traian made this change and the one from `applyRulesSequence`. @traiansf , could you help us? I couldn't follow you at this part

Comment (andreiburdusa) [reply kore/src/Kore/Step/Step.hs]: But I think that was moved to `applyRewriteRulesParallel`
https://github.com/kframework/kore/pull/1026/files/c5216f71d35ce543a37d6eba5c3d8384e546522c#diff-595068f6f4dcaabbeb534249f4fd4c12R620

File: kore/test/Test/Kore/Step/Step.hs (hunk 22 above)
Comment (andreiburdusa) [reply kore/test/Test/Kore/Step/Step.hs]: It's used with `applyRewriteRulesSequence` at line 347, but I will change it to make it look cleaner

File: src/main/k/working/imp/merge/2.merge-output [hunk 25]
```diff
@@ -0,0 +1,77 @@
+\rewrites{SortGeneratedTopCell{}}(
```
Comment (ttuegel) [inline src/main/k/working/imp/merge/2.merge-output]: This shouldn't be checked in. Could we remove it, and add a `.gitignore` rule for it?

File: src/main/kore/prelude.kore (hunk 23 above)
Comment (andreiburdusa) [reply src/main/kore/prelude.kore]: It was not. I removed it

File: kore/src/Kore/Unification/Unify.hs [hunk 26]
```diff
@@ -169,3 +173,8 @@ runUnifierT = runExceptT . BranchT.gather . getUnifierT
  -}
 maybeUnifierT :: MonadSimplify m => UnifierT m a -> MaybeT m [a]
 maybeUnifierT = hushT . BranchT.gather . getUnifierT
+
+{- | Run a 'Unifier', returning empty list upon error.
+ -}
+listUnifierT :: MonadSimplify m => UnifierT m a -> m [a]
```
[the remark this answers is not in the document]
Comment (andreiburdusa) [reply kore/src/Kore/Unification/Unify.hs]: I will remove the `T` from the name since it doesn't make sense. But why is it not safe? `fromRight` takes `[]` as default value

File: src/main/k/working/tests/fresh-vars/fresh-vars.k (hunk 24 above)
Comment (andreiburdusa) [reply src/main/k/working/tests/fresh-vars/fresh-vars.k]: this was useful for debugging. I changed it back

File: kore/src/Kore/Step/Step.hs [hunk 27]
```diff
@@ -409,6 +410,24 @@ finalizeRulesParallel initial unifiedRules = do
             $ Pattern.mapVariables Target.unwrapVariable <$> remainders'
         }
 
+checkFunctionLikeResults
```
[the remark this answers is not in the document]
Comment (andreiburdusa) [reply kore/src/Kore/Step/Step.hs]: I splitted `checkFunctionLikeResults` in two other functions, namely `assertFunctionLikeResults` and `recoveryFunctionLikeResults`. Maybe it's more intuitive now
